sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/CatalogExtensionForTableProvider.scala

```diff
@@ -0,0 +1,98 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.datasources.v2
+
+import java.util
+
+import scala.util.control.NonFatal
+
+import org.apache.spark.sql.AnalysisException
+import org.apache.spark.sql.connector.catalog.{DelegatingCatalogExtension, Identifier, SupportsSpecifiedSchemaPartitioning, Table}
+import org.apache.spark.sql.connector.expressions.Transform
+import org.apache.spark.sql.execution.datasources.DataSource
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.util.CaseInsensitiveStringMap
+
+class CatalogExtensionForTableProvider extends DelegatingCatalogExtension {
+
+  private val conf = SQLConf.get
+
+  override def loadTable(ident: Identifier): Table = {
+    val table = super.loadTable(ident)
+    tryResolveTableProvider(table)
+  }
+
+  override def createTable(
+      ident: Identifier,
+      schema: StructType,
+      partitions: Array[Transform],
+      properties: util.Map[String, String]): Table = {
+    val provider = properties.getOrDefault("provider", conf.defaultDataSourceName)
+    val maybeProvider = DataSource.lookupDataSourceV2(provider, conf)
+    val (actualSchema, actualPartitioning) = if (maybeProvider.isDefined && schema.isEmpty) {
+      // A sanity check. The parser should guarantee it.
+      assert(partitions.isEmpty)
+      // If `CREATE TABLE ... USING` does not specify table metadata, get the table metadata from
+      // data source first.
+      val table = maybeProvider.get.getTable(new CaseInsensitiveStringMap(properties))
+      table.schema() -> table.partitioning()
```

Comment:
This block isn't correct. If the table exists, then this must throw `TableAlreadyExistsException`: https://github.com/apache/spark/blob/master/sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/TableCatalog.java#L99

Create table is not allowed to return an existing table and override partitioning.

Reply:
There are 2 "table" in this context:
1. the table entry in Spark metastore
2. the table returned by `TableProvider`

For example, `CREATE TABLE abc USING jdbc OPTIONS(table='xyz')`, `abc` is the table entry in Spark metastore, `xyz` is the table returned by `TableProvider`. The table entry in Spark metastore is simply a link to the table returned by `TableProvider`. So the CREATE TABLE here is not to create a table in JDBC, but to create a table in Spark metastore that links to the JDBC table.

Reply:
Okay, so this create table uses the schema and partitioning from the external table because it wasn't specified in the DDL create statement. Sounds reasonable.

I don't think this implementation is correct. There are two cases: the schema and/or partitioning were not specified and are empty, or they were specified and need to be validated. I think that this should load the external table and either override (if empty) or validate if non-empty.

Looks like we didn't think of this case in the API, but we should add `IllegalArgumentException` for when the external table schema or partitioning doesn't match the incoming settings.

When the `inferSchema` and `inferPartitioning` methods are added, we will need to update this to use those methods instead of loading the table and getting its schema and partitioning. After those methods are added, the table implementation must always report the schema and partitioning that were passed in to build the table.